src/snowflake/snowpark/async_job.py

```diff
@@ -140,35 +171,72 @@ class AsyncJob:
     def __init__(
         self,
         query_id: str,
-        query: str,
-        server_connection: "snowflake.snowpark._internal.server_connection.ServerConnection",
-        data_type: _AsyncDataType = _AsyncDataType.ROW,
+        query: Optional[str],
+        session: "snowflake.snowpark.session.Session",
+        result_type: _AsyncResultType = _AsyncResultType.ROW,
+        post_actions: Optional[List[Query]] = None,
         **kwargs,
     ) -> None:
-        #: The query ID of the executed query
-        self.query_id: str = query_id
-        #: The SQL text of of the executed query
-        self.query: str = query
-        self._conn = server_connection._conn
-        self._cursor = self._conn.cursor()
-        self._data_type = data_type
+        self.query_id: str = query_id  #: The query ID of the executed query
+        self._query = query
+        self._can_query_be_retrieved = True
+        self._session = session
+        self._cursor = session._conn._conn.cursor()
+        self._result_type = result_type
+        self._post_actions = post_actions if post_actions else []
         self._parameters = kwargs
         self._result_meta = None
-        self._server_connection = server_connection
         self._inserted = False
         self._updated = False
         self._deleted = False
-        self._plan = None
+
+    @property
+    @experimental(version="0.10.0")
+    def query(self) -> Optional[str]:
```

Comment:
What does it return when multiple queries are grouped into a script?

Reply:
If it's a binding query (create_dataframe from large local data), it will be the last query (select from temp table). For other multiple queries, they will be wrapped into a snow script (anonymous block).